bundler/lib/dependabot/bundler/update_checker/latest_version_finder/dependency_source.rb

```diff
@@ -98,46 +103,53 @@ def dependency_rubygems_uri
 
           def private_registry_versions
             @private_registry_versions ||=
-              in_a_temporary_bundler_context do
-                bundler_source.
-                  fetchers.flat_map do |fetcher|
-                    fetcher.
-                      specs_with_retry([dependency.name], bundler_source).
-                      search_all(dependency.name)
-                  end.
-                  map(&:version)
+              in_a_native_bundler_context do |tmp_dir|
+                SharedHelpers.run_helper_subprocess(
+                  command: NativeHelpers.helper_path,
+                  function: "private_registry_versions",
+                  args: {
+                    dir: tmp_dir,
+                    gemfile_name: gemfile.name,
+                    dependency_name: dependency.name,
+                    credentials: credentials
+                  }
+                ).map do |version_string|
+                  Gem::Version.new(version_string)
+                end
               end
-          end
-
-          def bundler_source
-            return nil unless gemfile
-
-            @bundler_source ||=
-              in_a_temporary_bundler_context do
-                definition = ::Bundler::Definition.build(gemfile.name, nil, {})
+          rescue Dependabot::SharedHelpers::HelperSubprocessFailed => e
+            if e.message.match(BundlerErrorPatterns::MISSING_AUTH_REGEX)
+              source = Regexp.last_match(:source)
+              raise Dependabot::PrivateSourceAuthenticationFailure, source
+            elsif e.message.match(BundlerErrorPatterns::BAD_AUTH_REGEX)
+              source = Regexp.last_match(:source)
+              raise Dependabot::PrivateSourceAuthenticationFailure, source
+            elsif e.message.match(BundlerErrorPatterns::HTTP_ERR_REGEX)
+              source = Regexp.last_match(:source)
+              raise Dependabot::PrivateSourceTimedOut, source
+            end
 
-                specified_source =
-                  definition.dependencies.
-                  find { |dep| dep.name == dependency.name }&.source
+            # TODO: Add testing and further exception case handling
```

Comment:
Looks like this is the same handling as `handle_bundler_errors`, where you thinking pair on a refactor of this once this is merged into the bundler 2 spike?

Reply:
Yeah, I was planning to leave that for now and only rescue the _tested_ error cases - we can either pair on it or branch out and do that

Reply:
I think when I wrote this I was thinking part of the error handling refactor could be [preparatory](https://martinfowler.com/articles/preparatory-refactoring-example.html) and we could ship (most of it) into master